include/GafferOSL/OSLTextureReader.h

```diff
@@ -0,0 +1,78 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2017, Image Engine. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFEROSL_OSLTEXTUREREADER_H
+#define GAFFEROSL_OSLTEXTUREREADER_H
+
+#include "IECore/Reader.h"
+#include "IECore/ImagePrimitive.h"
+// #include "IECore/CompoundObject.h"
+
+#include "GafferOSL/TypeIds.h"
+
+namespace GafferOSL
+{
+
+class OSLTextureReader : public IECore::Reader
```

Comment:
Ingenious! Unless I'm missing something, what you really want to do is pass an ImagePrimitive directly as a shader parameter to OpenGL? That isn't possible because shader parameters have to be Data-derived, but there is already a workaround that allows a CompoundData with the right format to be passed. The OpenGLShader [already uses that](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/OpenGLShader.cpp#L174), so perhaps we could use the same here, and not need this OSLTextureReader class at all?